### ISSUE TITLE
Expect to have_received(method).with({}) fails when method modifies the Hash

### ISSUE THREAD

**Issue by bvrooman** [rspec/rspec-mocks#1251]
## What Ruby, Rails and RSpec versions are you using?

Ruby version: ruby 2.3.7p456 (2018-03-28 revision 63024) [x86_64-darwin17]
Rails version: Rails 5.2.1
Rspec version: RSpec 3.8

## Observed behaviour

I am using Rspec to test the presence of a method call with the correct parameters.
I invoke the method `call` with
```
my_callable.call({})
```
and I test the call using
```
expect(my_callable).to have_received(:call).with({})
```
When `call` modifies the provided hash argument, like `context[:value] = true`, this test fails, when it should succeed. This produces the output:
```
#<MyCallable:0x00007f9505aa5518> received :call with unexpected arguments
  expected: ({})
       got: ({:value=>true})
Diff:
@@ -1,2 +1,2 @@
-[{}]
+[{:value=>true}]
```

It appears that the `expect` call is referencing the hash modified in memory, rather than a saved copy of the hash when the call was actually made. 

## Expected behaviour

I expect this test to pass.
Because the call to the method `call` was made with an empty hash `{}`, the expect statement
```
expect(my_callable).to have_received(:call).with({})
```
is correct. 

## Can you provide an example app?

Link to repo:
https://github.com/bvrooman/Rspec-Issue-2044

I found that this contrived code can reproduce the issue:

Inside any `_spec` file:
```
class MyCallable
  def call(context)
    context[:value] = true
  end
end

let(:my_callable) { MyCallable.new }

describe 'MyCallable #call' do
  it 'is called with the right arguments' do
    allow(my_callable).to receive(:call).and_call_original
    my_callable.call({})
    expect(my_callable).to have_received(:call).with({})
  end
end
```

This results in the following failure:
```
#<MyCallable:0x00007f9505aa5518> received :call with unexpected arguments
  expected: ({})
       got: ({:value=>true})
Diff:
@@ -1,2 +1,2 @@
-[{}]
+[{:value=>true}]
```

**Comment by JonRowe** [comment]
Transferred to rspec/rspec-mocks as this is not an rspec-rails issue.

**Comment by benoittgt** [comment]
Thanks @bvrooman for the details bug report. I am working on your issue.

First. Did you know you can do:
```ruby
class MyCallable
  def call(context)
    context[:value] = true
  end
end


RSpec.describe 'additions' do
  let(:my_callable) { MyCallable.new }
  describe 'MyCallable #call' do
    it 'is called with the right arguments' do
      expect(my_callable).to receive(:call).with({})
      my_callable.call({})
    end
  end
end
```

And it works.

I will keep searching for your code example and let you know what I found

**Comment by benoittgt** [comment]
This is a WIP comment. I dig into this subject in train but didn't found a solution.

----

It will also work if you remove `.and_call_original`.

Maybe their an issue with Mutex and [how we read arg value](https://vaneyckt.io/posts/ruby_concurrency_in_praise_of_the_mutex/).

```
From: /Users/benoit.tigeot/code/rspec-dev/repos/rspec-mocks/lib/rspec/mocks/message_expectation.rb @ line 577 :

    572:           end
    573:
    574:           @order_group.handle_order_constraint self
    575:
    576:           if implementation.present?
 => 577:             binding.irb
    578:             implementation.call(*args, &block)
    579:           elsif parent_stub
    580:             parent_stub.invoke(nil, *args, &block)
    581:           end
    582:         ensure

2.5.1 :001 > args
 => [{}]
2.5.1 :002 > implementation.call(*args, &block)
 => true
2.5.1 :003 > args
 => [{:value=>true}]
```

This part is quite new to me and I am not good at debugging multithreaded issue if it one.
I think we do:
1. Spie the method for a call with `{}`. It works ok.
2. We call `.and_call_original` but `args`get mutated
3. We check for message expectation and see mutated argument

**Comment by bvrooman** [comment]
Hi @benoittgt , 

Thanks for investigating the issue. 

We are not blocked by this issue, so it is a low priority for us. We can definitely use one of the alternatives that you propose in the mean time. 

Please let me know if there are any other details I can provide for your investigation. If I find some time, I can dig deeper into it too, following the research you've done.

**Comment by BobbyMcWho** [comment]
It seems like this is [documented behavior](https://github.com/rspec/rspec-mocks/blob/7dc2a431101172d31b780458eb168e05d20fb3aa/lib/rspec/mocks/example_methods.rb#L279), is this a behavior that warrants fixing, or should the alternative expect before calling be the recommended usage?

**Comment by JonRowe** [comment]
If it can be fixed it should

**Comment by benoittgt** [comment]
The problem is clearer now. I think spies could store the arguments when they are called. So the expectation could be made against the stored calls and not when the example is ending (with argument reference mutated).

We could at first trigger a warning.

**Comment by williamhakim10** [comment]
@benoittgt thanks for providing some ideas :)
I definitely think this is worth considering, because while it makes sense how this works once you wrap your head around it, it's not the behavior I would have intuitively expected and took a long time to debug before I figured out what was actually going on.

**Comment by Diogoazvg** [comment]
I think your class should be instantiated first, have the call method permission in case you make use of have_received expectation.
The two ways below work with both the prior and the posterior expectation.

```ruby
class MyCallable < Object
  def call(context)
    context[:value] = true
  end
end

RSpec.describe 'MyCallable #call' do
  let(:my_callable_instance) { MyCallable.new }
  let(:my_callable) { instance_double(MyCallable) }

  it 'is called with the right arguments' do
    allow(MyCallable).to receive(:new) { my_callable }
    allow(my_callable).to receive(:call)

    my_callable_instance.call({})
    
    expect(my_callable).to have_received(:call).with({})
  end

  it 'is called with the right arguments 2' do
    allow(MyCallable).to receive(:new) { my_callable }
    
    expect(my_callable).to receive(:call).with({})
    
    my_callable_instance.call({})
  end
end
```
